### PR TITLE
Add amp -ui

### PR DESCRIPTION
**Pull request by freignat91**
add amp-ui, client Angular 4, server Go, first version

## test

- ampmake clean build
- amp cluster create --server localhost --tag local

Once the cluster is up, we have for now to launch manually the infrastructure ampui.stack.yml stack which now contains amp-ui:

    $amp stack deploy -c stacks/ampui.stack.yml ui

- then open Chrome and browse url:  https://ui.local.atomiq.io

image

image

### CODE REVIEW

File: amp-ui/client/src/app/services/stacks.service.ts [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+
+@Injectable()
+export class StacksService {
+  stacks : Stack[] = []
+  stack : Stack = null
+  onStacksLoaded = new Subject();
+  onStacksError = new Subject();
+
+  constructor(private httpService : HttpService) {
+    this.stacks.push(new Stack("907345782", "haproxy", 1, "freignat", "USER"))
```
Comment (subfuzion) [inline amp-ui/client/src/app/services/stacks.service.ts:14]: @freignat91 What's all this?

File: glide.lock [hunk 2]
```diff
@@ -1,5 +1,5 @@
 hash: fff401d3955c816110993c628fa41dd5f983036587ee137479f478b3e967dccb
-updated: 2017-04-26T21:39:21.015444832Z
+updated: 2017-04-26T23:59:00.032458759Z
```
Comment (subfuzion) [inline glide.lock:2]: Glide update shouldn't be included in PR.

File: amp-ui/client/src/app/services/stacks.service.ts (hunk 1 above)
Comment (freignat91) [reply amp-ui/client/src/app/services/stacks.service.ts:14]: It's a hardcoded stack i used for test, to be removed

File: glide.lock (hunk 2 above)
Comment (freignat91) [reply glide.lock:2]: ok, i should be able to remove this vendor update amp-ui doesn't need

Comment (subfuzion) [reply glide.lock:2]: Ok, but I thought that's what the vendor update from yesterday was for?

**Review by subfuzion — CHANGES_REQUESTED** [review]
@freignat91 Great work overall! I'll be honest, I'm on the fence on one thing, however: one advantage of this particular gateway approach is we can point to any amp server (although I don't see this as a big use case right now); but a big disadvantage is that we now need to manage a separate server, and manually forward/transform what was automatically mapped for us using grpc REST gateway support. This approach creates more build artifacts, duplicates code and creates more code paths to be tested, and doesn't exploit available automated code generation that's generally preferable over manual effort. This is something we will want to discuss and resolve before we go too far down a particular path and the codebase for this blossoms. Let's take time for an offline discussion over this. The path I really prefer is to support auto-generation for multiple grpc clients, and then (and only if really necessary) build a lightweight JavaScript client library that simplifies any remaining logic for establishing a connection, setting any custom HTTP headers for authentication tokens, etc, that make it easy to call the backend.

**Comment by subfuzion** [comment]
@freignat91 Updating the description `docker run --rm -v $PWD/stacks:/stacks:ro --network hostnet docker -H m1 stack deploy -c stacks/ampui.stack.yml amp` -> `amp stack deploy -c stacks/ampui.stack.yml ui`.

**Review by subfuzion — APPROVED** [review]

**Comment by subfuzion** [comment]
Nice work, @freignat91. Approving as is and we'll continue the discussion about alternatively restoring the REST gateway support over the next day or so. 👍

**Pull request merged**